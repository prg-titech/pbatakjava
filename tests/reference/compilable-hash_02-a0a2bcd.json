{
    "basename": "compilable-hash_02-a0a2bcd",
    "cmd": "{infile}",
    "infile": "tests/hash_02/build.sh",
    "infile_hash": "fec4d6709bfd9781f16fb290dee0dc55eea88f36e670def985d0e3a5",
    "stdout": "compilable-hash_02-a0a2bcd.stdout",
    "stdout_hash": "3a8951ca76fd03cc677f5f60e68cd9f5007818ee0831d5fd4bb80166",
    "stderr": null,
    "stderr_hash": null,
    "returncode": 0
}