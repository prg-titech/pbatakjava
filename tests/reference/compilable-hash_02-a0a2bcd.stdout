
public interface HashValue {
  public int getVal();

  public boolean equals(HashValue o);
}

public interface HashValue_Factory {
  public HashValue make(int val);
}

public class HashValue_v2 implements HashValue {
  private int val;

  public HashValue_v2(int val) {
    this.val = val;
  }

  public int getVal() {
    return this.val;
  }

  public boolean equals(HashValue o) {
    return this.getVal() == o.getVal();
  }
}

public class HashValue_v2_Factory implements HashValue_Factory {
  public HashValue make(int val) {
    return new HashValue_v2(val);
  }
}

public class HashValue_v1 implements HashValue {
  private int val;

  public HashValue_v1(int val) {
    this.val = val;
  }

  public int getVal() {
    return this.val;
  }

  public boolean equals(HashValue o) {
    return this.getVal() == o.getVal();
  }
}

public class HashValue_v1_Factory implements HashValue_Factory {
  public HashValue make(int val) {
    return new HashValue_v1(val);
  }
}

public interface Hash {
  public HashValue mkHash(String s);

  public boolean match(String s, HashValue h);
}

public interface Hash_Factory {
  public Hash make();
}

public class Hash_v1 implements Hash {
  public Hash_v1() {
  }

  public HashValue mkHash(String s) {
    if (s.equals("a")) {
      return new HashValue_v1(1);
    }
    return new HashValue_v1(0);
  }

  public boolean match(String s, HashValue h) {
    return this.mkHash(s).equals(h);
  }
}

public class Hash_v1_Factory implements Hash_Factory {
  public Hash make() {
    return new Hash_v1();
  }
}

public class Hash_v2 implements Hash {
  public Hash_v2() {
  }

  public HashValue mkHash(String s) {
    if (s.equals("a")) {
      return new HashValue_v2(-1);
    }
    return new HashValue_v2(0);
  }

  public boolean match(String s, HashValue h) {
    return this.mkHash(s).equals(h);
  }
}

public class Hash_v2_Factory implements Hash_Factory {
  public Hash make() {
    return new Hash_v2();
  }
}

public interface Dir {
  public String[] getFiles();

  public boolean exists(HashValue h);
}

public interface Dir_Factory {
  public Dir make(HashValue_Factory V, Hash_Factory W);
}

public class Dir_v1 implements Dir {
  public HashValue_Factory V;

  public Hash_Factory W;

  public Dir_v1(HashValue_Factory V, Hash_Factory W) {
    this.V = V;
    this.W = W;
  }

  public String[] getFiles() {
    String[] files = new String[1];
    files[0] = "a";
    return files;
  }

  public boolean exists(HashValue h) {
    String[] fs = this.getFiles();
    Hash hash = W.make();
    for (int i = 0; i < fs.length; i++) {
      if (hash.match(fs[i], h)) {
        return true;
      }
    }
    return false;
  }
}

public class Dir_v1_Factory implements Dir_Factory {
  public Dir make(HashValue_Factory V, Hash_Factory W) {
    return new Dir_v1(V, W);
  }
}

public class Test {
  public static void main(String[] args) {
    String s = args[0];
    Hash hash1 = new Hash_v1();
    HashValue digest1 = hash1.mkHash(s);
    Hash hash2 = new Hash_v2();
    HashValue digest2 = hash2.mkHash(s);
    Dir dir1 = new Dir_v1(new HashValue_v1_Factory(), new Hash_v1_Factory());
    Dir dir2 = new Dir_v1(new HashValue_v2_Factory(), new Hash_v2_Factory());
    System.out.println(dir1.exists(digest1));
    System.out.println(dir2.exists(digest2));
  }
}
